blockchain/src/codec.rs

```diff
@@ -0,0 +1,376 @@
+use readerwriter::{Decodable, Encodable, ReadError, Reader, WriteError, Writer};
+use std::convert::TryFrom;
+use zkvm::{merkle, Hash, Signature, Tx};
+
+#[repr(u8)]
+enum MessageType {
+    Block = 0,
+    GetBlock = 1,
+    Inventory = 2,
+    GetInventory = 3,
+    MempoolTxs = 4,
+    GetMempoolTxs = 5,
+}
+
+impl TryFrom<u8> for MessageType {
+    type Error = ReadError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MessageType::Block),
+            1 => Ok(MessageType::GetBlock),
+            2 => Ok(MessageType::Inventory),
+            3 => Ok(MessageType::GetInventory),
+            4 => Ok(MessageType::MempoolTxs),
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+fn read<T>(data: Result<T, ReadError>, label: &'static str) -> Result<T, ReadError> {
+    data.map_err(|err| {
+        ReadError::Custom(
+            match err {
+                ReadError::InvalidFormat => format!("invalid format of {}", label),
+                r => r.to_string(),
+            }
+            .into(),
+        )
+    })
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
+        let s = self.read_u8x32()?;
+        let r = self.read_u8x32()?;
+        Ok(Signature {
+            s: Scalar::from_bits(s),
+            R: CompressedRistretto(r),
+        })
+    }
+
+    fn read_shortid_vec(&mut self) -> Result<ShortIDVec, ReadError> {
+        self.read_u8_vec().map(ShortIDVec)
+    }
+
+    fn read_blockid(&mut self) -> Result<BlockID, ReadError> {
+        self.read_u8x32().map(BlockID)
+    }
+
+    fn read_hash(&mut self) -> Result<Hash, ReadError> {
+        self.read_u8x32().map(Hash)
+    }
+
+    fn read_block_header(&mut self) -> Result<BlockHeader, ReadError> {
+        Ok(BlockHeader {
+            version: self.read_u64()?,
+            height: self.read_u64()?,
+            prev: self.read_blockid()?,
+            timestamp_ms: self.read_u64()?,
+            txroot: self.read_hash()?,
+            utxoroot: self.read_hash()?,
+            ext: self.read_u8_vec()?,
+        })
+    }
+
+    fn read_inventory(&mut self) -> Result<Inventory, ReadError> {
+        Ok(Inventory {
+            version: self.read_u64()?,
+            tip: self.read_block_header()?,
+            tip_signature: self.read_signature()?,
+            shortid_nonce: self.read_u64()?,
+            shortid_list: self.read_shortid_vec()?,
+        })
+    }
+
+    fn read_tx(&mut self) -> Result<Tx, ReadError> {
+        let tx_size = self.read_u32()? as usize;
+        let vec = self.read_vec(tx_size)?;
+        Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
```

Comment:
Maybe `impl Decodable` for `zkvm::Tx`?

Reply:
Moved these functions outside of ReaderExt, because Tx implement Encodable already.